src/Stratis.Bitcoin.Features.BlockStore.Tests/BlockStoreTests.cs

```diff
@@ -0,0 +1,241 @@
+            int blockSize = block.GetSerializedSize();
+
+            int count = (5 * 1000 * 1000) / blockSize + 2;
+
+            ConcurrentChain longChain = this.CreateChain(count);
+            this.consensusTip = longChain.Tip;
+            this.repository.BlockHash = longChain.Genesis.HashBlock;
+
+            this.blockStore = new BlockStore(this.repository, longChain, this.chainState, new StoreSettings(), new NodeLifetime(), new LoggerFactory());
+
+            await this.blockStore.InitializeAsync().ConfigureAwait(false);
+
+            for (int i = 1; i < count; i++)
+            {
+                ChainedHeader header = longChain.GetBlock(i);
+
+                this.blockStore.AddToPending(new BlockPair(block, header));
+            }
+            
+            await this.WaitUntilQueueIsEmpty().ConfigureAwait(false);
+            Assert.Equal(this.blockStore.StoreTip, longChain.GetBlock(count - 1));
+            Assert.Equal(1, this.repository.SavesCount);
+        }
+
+        [Fact]
+        public async Task BatchIsSavedOnShutdownAsync()
+        {
+            this.repository.BlockHash = this.chain.Genesis.HashBlock;
+            
+            await this.blockStore.InitializeAsync().ConfigureAwait(false);
+
+            ChainedHeader lastHeader = null;
+
+            for (int i = 1; i < this.chain.Height - 1; i++)
+            {
+                lastHeader = this.chain.GetBlock(i);
+                this.blockStore.AddToPending(new BlockPair(new Block(), lastHeader));
+            }
+
+            await this.WaitUntilQueueIsEmpty().ConfigureAwait(false);
+
+            Assert.Equal(this.blockStore.StoreTip, this.chain.Genesis);
+            Assert.Equal(0, this.repository.SavesCount);
+            
+            this.nodeLifetime.StopApplication();
+            this.blockStore.Dispose();
+
+            Assert.Equal(this.blockStore.StoreTip, lastHeader);
+            Assert.Equal(1, this.repository.SavesCount);
+        }
+
+        [Fact]
+        public async Task BatchIsSavedWhenAtTheTipAsync()
+        {
+            this.repository.BlockHash = this.chain.Genesis.HashBlock;
+            this.consensusTip = this.chain.Tip;
+
+            await this.blockStore.InitializeAsync().ConfigureAwait(false);
+
+            for (int i = 1; i <= this.chain.Height; i++)
+            {
+                ChainedHeader lastHeader = this.chain.GetBlock(i);
+                this.blockStore.AddToPending(new BlockPair(new Block(), lastHeader));
+            }
+
+            await this.WaitUntilQueueIsEmpty().ConfigureAwait(false);
+            
+            Assert.Equal(this.blockStore.StoreTip, this.chain.Tip);
+            Assert.Equal(1, this.repository.SavesCount);
+        }
+    }
+
+    public class BlockRepositoryMock : IBlockRepository
+    {
+        /// <summary><see cref="GetAsync"/> will return blocks from this list.</summary>
```

Comment:
I can't see the connection between this summary and the line below

Reply:
It says that GetAsync will return blocks from that list, not from the database (since the class is a mock and we are imitating the behavior)

Reply:
ah right, so it is really  `blockRepository.GetAsync` that is setup to return it. I guess if you have switched it to moq then that is probably more clear already in the next commit.

Reply:
Sure. Moved it to a variable